allennlp/training/metrics.py

```diff
@@ -0,0 +1,185 @@
+from typing import Dict, Optional
+
+from overrides import overrides
+import torch
+
+from allennlp.common.registrable import Registrable
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.params import Params
+
+
+class Metric(Registrable):
+    """
+    A very general abstract class representing a metric which can be
+    accumulated.
+    """
+    def __call__(self, predictions: torch.Tensor,
+                 gold_labels: torch.Tensor,
+                 mask: Optional[torch.Tensor]):
+
+        raise NotImplementedError
+
+    def get_metric(self, reset: bool) -> Dict[str, float]:
```

Comment:
Well, I see that your `F1` metric returns three things, so now I get why you did it this way.  I think it'd be a bit cleaner, though, if this returned `Union[float, Tuple[float, ...]]`, and the mapping to strings still happened in the model.  What if you want to compute several metrics, for instance?  How are you supposed to reconcile naming conflicts in the metrics you're calling?  Much simpler if the model gives them names, instead of the metrics themselves.